### PR TITLE
Add new plugin metadata and hooks for starting only Processing component of plugins

### PR DESCRIPTION
**Pull request by nyalldawson**
Allows for identification of plugins which implement Processing functionality (without having to load them first), and the new initProcessing() hook allows for plugins to load only their Processing functionality (without any GUI or iface dependant components)

### CODE REVIEW

File: tests/src/app/testqgisapppython.cpp [hunk 1]
```diff
@@ -73,6 +79,58 @@ void TestQgisAppPython::cleanupTestCase()
...
+  QVERIFY( mQgisApp->mPythonUtils->listActivePlugins().isEmpty() );
+  // load plugin
+  QVERIFY( !mQgisApp->mPythonUtils->unloadPlugin( QStringLiteral( "PluginPathTest" ) ) );
+  QVERIFY( mQgisApp->mPythonUtils->loadPlugin( QStringLiteral( "PluginPathTest" ) ) );
+  QVERIFY( !mQgisApp->mPythonUtils->isPluginLoaded( QStringLiteral( "PluginPathTest" ) ) );
+  QVERIFY( mQgisApp->mPythonUtils->startPlugin( QStringLiteral( "PluginPathTest" ) ) );
+  QVERIFY( mQgisApp->mPythonUtils->isPluginLoaded( QStringLiteral( "PluginPathTest" ) ) );
+  QCOMPARE( mQgisApp->mPythonUtils->listActivePlugins(), QStringList() << QStringLiteral( "PluginPathTest" ) );
+
+#if 0 // unreliable on Travis?
```
Comment (m-kuhn) [inline tests/src/app/testqgisapppython.cpp]: Any context to that?

File: src/python/qgspythonutilsimpl.cpp [hunk 2]
```diff
@@ -623,6 +628,13 @@ bool QgsPythonUtilsImpl::startPlugin( const QString &packageName )
   return ( output == QLatin1String( "True" ) );
 }
 
+bool QgsPythonUtilsImpl::startProcessingPlugin( const QString &packageName )
+{
+  QString output;
+  evalString( "qgis.utils.startProcessingPlugin('" + packageName + "')", output );
```
Comment (m-kuhn) [inline src/python/qgspythonutilsimpl.cpp]:
```suggestion
  evalString( QStringLiteral( "qgis.utils.startProcessingPlugin('" ) + packageName + QStringLiteral( "')" ), output );
```

**Comment by m-kuhn** [comment]
Good idea. +1 for a test ride!

**Review by m-kuhn — APPROVED** [review]

File: tests/src/app/testqgisapppython.cpp (hunk 1 above)
Comment (nyalldawson) [reply tests/src/app/testqgisapppython.cpp]: The call to unloadPlugin works locally, but fails on Travis. I could see this quickly becoming one of those Travis black holes which eats up days, and I didn't consider it worth the effort given that plugin unload works stably and has for xxx years. I'll just remove the whole block instead.

**Comment by Gustry** [comment]
This is a nice idea!
This needs some documentation: QGIS Plugin Builder, QGIS Cookbook ...

No label `[needs-docs]` was added. Are you waiting a few weeks first for a test?

We can encourage the use of Processing within plugin now: bringing plugins on the foreground on plugins.qgis.org which implement Processing provider etc...

File: src/python/qgspythonutilsimpl.cpp [hunk 3]
```diff
@@ -603,49 +603,61 @@ QStringList QgsPythonUtilsImpl::pluginList()
...
-  QString str = "qgis.utils.pluginMetadata('" + pluginName + "', '" + function + "')";
+  QString str = QStringLiteral( "qgis.utils.pluginMetadata('%1', '%2')" ).arg( pluginName, function );
   evalString( str, res );
   //QgsDebugMsg("metadata "+pluginName+" - '"+function+"' = "+res);
   return res;
 }
 
+bool QgsPythonUtilsImpl::pluginHasProcessingProvider( const QString &pluginName )
+{
+  return getPluginMetadata( pluginName, QStringLiteral( "hasProcessingProvider" ) ).compare( QLatin1String( "yes" ), Qt::CaseInsensitive ) == 0;
```
Comment (Gustry) [inline src/python/qgspythonutilsimpl.cpp]: @nyalldawson I just noticed that in the `metadata.txt`, we have booleans for `experimental`, `deprecated` and `server`.

Your new parameter `hasProcessingProvider` is a `yes`/`no`. Can we change?

**Pull request merged**